### PR TITLE
Major refactor of cover platform and config flow support tweaks for cover platform

### PR DESCRIPTION
**Pull request by andrewmeepos**
I'm bad at GitHub and didn't know how to update my old PR to run on top of the new master. Therefore, i'm closing https://github.com/rospogrigio/localtuya-homeassistant/pull/10 and opening this new PR. I'm now getting the following error when setting up a cover from the config flow:

```
2020-09-16 01:17:29 ERROR (MainThread) [homeassistant.components.cover] Error while setting up localtuya platform for cover
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 184, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/localtuya/cover.py", line 111, in async_setup_entry
    TuyaCache(device, config_entry.data[CONF_FRIENDLY_NAME]),
KeyError: 'friendly_name'`
```

@postlund @rospogrigio can either of you help me debug? My efforts to fix the error have thus far been unfruitful.

### CODE REVIEW

**Comment by rospogrigio** [comment]
I'm getting the same error so that's not your fault. Will try to fix it when I get the time.

**Comment by postlund** [comment]
This is because of the changes in #12 regarding `friendly_name` that should be reverted.

**Comment by postlund** [comment]
Now I kinda suggest that we revert #12, get #18 in place and fix the same thing for the other platforms (as in #18). Then we get a lot of code-reuse that simplifies `device_info`.

**Comment by rospogrigio** [comment]
> This is because of the changes in #12 regarding `friendly_name` that should be reverted.

> Now I kinda suggest that we revert #12, get #18 in place and fix the same thing for the other platforms (as in #18). Then we get a lot of code-reuse that simplifies `device_info`.

@postlund can I ask you to do these changes in a new PR? Because I'm not sure I get what you mean, in particular for the `friendly_name` thing. I am still working on #17, in the meantime.

**Comment by postlund** [comment]
@rospogrigio I can try to fix it later.

**Comment by postlund** [comment]
I believe I managed to revert the changes I meant in #20, but I'm not entirely sure. It's really, really hard to follow with all the crazy merges left and right. It would help a lot if you stopped using merge and did rebase (with `push -f`) instead. It would keep the history a lot cleaner.

**Comment by rospogrigio** [comment]
😢 sorry but can you provide the correct commands I should use? I have been using the commands suggested here on GitHub:
```
git checkout master
git merge --no-ff postlund-fix_friendly_name
git push origin master
```
I'll try it on #17 that is ready for merging so you can tell me if I'm doing it right. Sorry for the mess, I am still a bit new to GitHub...

**Comment by postlund** [comment]
No worries, it takes some time to get used to but is so much better in the end. I think the first "Working Locally" explains it a bit:

https://gist.github.com/markreid/12e7c2203916b93d23c27a263f6091a0

**Comment by rospogrigio** [comment]
I followed the instructions but I am still staying in my local branch, master is not being updated... what am I missing?
Here is what I tried:
```
root@rospberrypi:~/etc/localtuya-homeassistant/custom_components/localtuya [branch postlund-config_flow_0d]# git commit -am "Introduced dps detection with bruteforcing for type_0d devices"
[postlund-config_flow_0d af77e4b] Introduced dps detection with bruteforcing for type_0d devices
 2 files changed, 54 insertions(+), 4 deletions(-)

root@rospberrypi:~/etc/localtuya-homeassistant/custom_components/localtuya [branch postlund-config_flow_0d]# git rebase origin/master
First, rewinding head to replay your work on top of it...
Applying: Request DPS 1-150 in config flow
Using index info to reconstruct a base tree...
M       custom_components/localtuya/config_flow.py
Falling back to patching base and 3-way merge...
Auto-merging custom_components/localtuya/config_flow.py
Applying: Introduced dps detection with bruteforcing for type_0d devices
root@rospberrypi:~/etc/localtuya-homeassistant/custom_components/localtuya [branch postlund-config_flow_0d]# git push -f
fatal: Il branch corrente postlund-config_flow_0d non ha alcun branch upstream.
Per eseguire il push del branch corrente ed impostare remote come upstream, usa
    git push --set-upstream origin postlund-config_flow_0d

root@rospberrypi:~/etc/localtuya-homeassistant/custom_components/localtuya [branch postlund-config_flow_0d]#  git push --set-upstream origin postlund-config_flow_0d
Enumerating objects: 18, done.
Counting objects: 100% (18/18), done.
Delta compression using up to 4 threads
Compressing objects: 100% (9/9), done.
Writing objects: 100% (12/12), 1.69 KiB | 133.00 KiB/s, done.
Total 12 (delta 7), reused 0 (delta 0)
remote: Resolving deltas: 100% (7/7), completed with 4 local objects.
remote:
remote: Create a pull request for 'postlund-config_flow_0d' on GitHub by visiting:
remote:      https://github.com/rospogrigio/localtuya-homeassistant/pull/new/postlund-config_flow_0d
remote:
To github.com:rospogrigio/localtuya-homeassistant.git
 * [new branch]      postlund-config_flow_0d -> postlund-config_flow_0d
Branch 'postlund-config_flow_0d' set up to track remote branch 'postlund-config_flow_0d' from 'origin'.

root@rospberrypi:~/etc/localtuya-homeassistant/custom_components/localtuya [branch postlund-config_flow_0d]# git fetch
root@rospberrypi:~/etc/localtuya-homeassistant/custom_components/localtuya [branch postlund-config_flow_0d]# git rebase origin/master
Current branch postlund-config_flow_0d is up to date.
root@rospberrypi:~/etc/localtuya-homeassistant/custom_components/localtuya [branch postlund-config_flow_0d]# git rebase master
Current branch postlund-config_flow_0d is up to date.
root@rospberrypi:~/etc/localtuya-homeassistant/custom_components/localtuya [branch postlund-config_flow_0d]#  git push -f
Everything up-to-date
```

**Comment by postlund** [comment]
At a quick glance it looks fine. `master` is not supposed to change, the branch must be merged first. You *always* push a local branch with changes, then create a PR, review it and when everything is done you merge it from GitHub. Never manually. There are different strategies for merging, which you find in the settings page. I suggest that you change to only allow rebase. Once the PR has been merged, you can fetch and rebase OR pull (same thing here):

```
$ git checkout master
$ git fetch origin
$ git rebase origin/master

OR

$ git pull --rebase
```

The whole point here is that your local branch should always only contain your commits on top of what is currently on `master`. So no merge commits or other strange stuff. When you rebase (to catch up), you break history and that is why you need to force push. You can no longer apply your changes on top of what is on GitHub.

**Comment by rospogrigio** [comment]
OK but how am I supposed to create a PR now?? I cannot find the remote branch if I make a New PR... is it because I kept the same branch name as yours? But I don't see my commits in #17 either...

**Comment by postlund** [comment]
I'm not sure either where your branch went. I usually just push it manually, you can try that:

```
$ git push -f origin HEAD:<your branch name>
```

Pick a new branch name and don't reuse branch names, in case you don't remove them on GitHub first (after merging). There's an option in the settings page to delete branches automatically after being merged, so you don't have to think about it. It's quite convenient so I suggest you enable that as well.

**Comment by rospogrigio** [comment]
OK probably it was because of that. Please check if I made it right with #21 and give me feedback... thank you
